### PR TITLE
populate a new mz_source_info system table with per partition ingest metrics

### PR DESCRIPTION
**Pull request by elindsey**
Related to https://github.com/MaterializeInc/materialize/issues/4104 , this records our partition offset received/ingested metrics that were previously only surfaced in prometheus to a system table as well.

It does an unfortunate amount of "threading the logger through" and duplication between sources; going to clean that up as we start to refactor the sources code.

There are two things missing here:
1. It doesn't track the max offset. Right now we get this from a separate polling thread off in the coord, so it'll take a bit of work to join it to these offsets coming from the workers
2. This is a somewhat shallow notion of progress, it needs to be paired with timestamps/frontiers to really answer "is the dataflow keeping up." That's in a separate task though.

Example showing metric population and drop on source removal:
```
materialize=> select * from mz_source_info;
      source_name      | source_id | partition_id | received_offset | ingested_offset 
-----------------------+-----------+--------------+-----------------+-----------------
 mysql.simple.user     | u3/18     | 0            |           72724 |           72724
 mysql.simple.region   | u2/18     | 0            |           13010 |           13010
 mysql.simple.purchase | u1/18     | 0            |           11281 |           11281
(3 rows)

materialize=> select * from mz_source_info;
      source_name      | source_id | partition_id | received_offset | ingested_offset 
-----------------------+-----------+--------------+-----------------+-----------------
 mysql.simple.user     | u3/18     | 0            |          112281 |          112281
 mysql.simple.region   | u2/18     | 0            |           13030 |           13030
 mysql.simple.purchase | u1/18     | 0            |           11291 |           11291
(3 rows)
drop view purchase_sum_by_region;
materialize=> select * from mz_source_info;
 source_name | source_id | partition_id | received_offset | ingested_offset 
-------------+-----------+--------------+-----------------+-----------------
(0 rows)
```

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4572)

### CODE REVIEW

**Comment by benesch** [comment]
Just driving by, but I think you could pull the logger out of https://docs.rs/timely/0.11.1/timely/worker/struct.Worker.html#method.log_register anywhere that you have a scope, if you wanted to avoid the plumbing. Not obviously better, but an option!

File: src/dataflow/src/source/mod.rs [hunk 1]
```diff
@@ -730,11 +734,52 @@ pub struct PartitionMetrics {
...
+    }
+
+    /// Record the latest offset ingested high-water mark
+    pub fn record_offset_ingested(&mut self, offset: i64) {
+        self.offset_ingested.set(offset);
+        if let Some(logger) = self.logger.as_mut() {
+            logger.log(MaterializedEvent::SourceInfo(
+                self.source_name.clone(),
+                self.source_id.clone(),
+                self.partition_id.clone(),
```
Comment (quodlibetor) [inline src/dataflow/src/source/mod.rs:768]: this is going to cause three allocations on every kafka batch, which is hot enough that we refactored to the current PartitionMetrics struct to relieve pressure on the allocator.

I'm not sure if it's possible for us to turn this into `logger.log(Arc<SourceInfo>)` or something similar? Possibly otherwise we should use [`smol_str`](https://github.com/rust-analyzer/smol_str) here, if that's doable inside of timely.

**Review by benesch — COMMENTED** [review]
> Reviewed 3 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4572)* status: 3 of 10 files reviewed, 7 unresolved discussions (waiting on @elindsey and @quodlibetor)

___
*[src/dataflow/src/logging/materialized.rs, line 44 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKGkIX67Nofinf5Stl1:-MKGkIX67Nofinf5Stl2:b3bk5gi) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/logging/materialized.rs#L44)):*
> ```Rust
>     Peek(Peek, bool),
>     /// Tracks the source name, id, partition id, and received/ingested offsets
>     SourceInfo(String, String, String, i64, i64),
> ```

I know you're just following what seems to be convention here, but tuple structs with this many parameters are pretty hard to follow. What about this?

```
SourceInfo {
    source_name: String,
    source_id: String,
    partition_id: String,
    received_offset: i64,
    ingested_offset: i64,
}
```
___
*[src/dataflow/src/source/mod.rs, line 768 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoh-r1-768:-MKGlmhPAs4n8PCEA0Bu:ba9fntd) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/source/mod.rs#L768)):*
<details><summary><i>Previously, quodlibetor (Brandon W Maister) wrote…</i></summary><blockquote>

this is going to cause three allocations on every kafka batch, which is hot enough that we refactored to the current PartitionMetrics struct to relieve pressure on the allocator.

I'm not sure if it's possible for us to turn this into `logger.log(Arc<SourceInfo>)` or something similar? Possibly otherwise we should use [`smol_str`](https://github.com/rust-analyzer/smol_str) here, if that's doable inside of timely.
</blockquote></details>

I'm not sure it's worth trying to avoid the allocations here, because the logger is going to cause another allocation when it packs these things into a row. But (if my understanding of how this code is called is correct) I agree that this loop is hot enough that we shouldn't allocate here.

Can we just refactor the Kafka code so that we only log the new offsets when we're done with a whole batch of messages? It seems like as written we'll log on every new message.

**Review by quodlibetor — DISMISSED** [review]
> Reviewed 1 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4572)* status: 4 of 10 files reviewed, 7 unresolved discussions (waiting on @benesch and @elindsey)

___
*[src/dataflow/src/source/mod.rs, line 768 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoh-r1-768:-MKGp36kA9JTTKKYlGgz:bxw2x3i) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/source/mod.rs#L768)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I'm not sure it's worth trying to avoid the allocations here, because the logger is going to cause another allocation when it packs these things into a row. But (if my understanding of how this code is called is correct) I agree that this loop is hot enough that we shouldn't allocate here.

Can we just refactor the Kafka code so that we only log the new offsets when we're done with a whole batch of messages? It seems like as written we'll log on every new message.
</blockquote></details>

4 tiny allocations are worse than 1, though, aren't they?

I thought that the rowpacker had an internal freelist of rows that it could use for its allocations, reducing the likelihood of going all the way out to malloc.

I agree that moving this out so that it's not done in a batch is important enough to do either way.

**Review by elindsey — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4572)* status: 4 of 10 files reviewed, 7 unresolved discussions (waiting on @elindsey and @quodlibetor)

___
*[src/coord/src/catalog/builtin.rs, line 98 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoa-r1-98:-MKGx9mL-Tonj8GI6r6s:brz6jmt) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/coord/src/catalog/builtin.rs#L98)):*
<details><summary><i>Previously, quodlibetor (Brandon W Maister) wrote…</i></summary><blockquote>

typo: this went from `makes the makes the` to `makes the the`
</blockquote></details>

Well, fixed half the original typo 🙂 
___
*[src/dataflow/src/logging/materialized.rs, line 258 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoc-r1-258:-MKGzvCSC5DXzIp5xJoc:ba0txnj) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/logging/materialized.rs#L258)):*
<details><summary><i>Previously, quodlibetor (Brandon W Maister) wrote…</i></summary><blockquote>

What does `count` do here? I'm unfamiliar with it.
</blockquote></details>

Count is summing our composite (offset_received_diff, offset_ingested_diff) value into an (offset_received_sum, offset_ingested_sum), adding up all the diffs for both to get the final offsets
___
*[src/dataflow/src/render/mod.rs, line 207 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceod-r1-207:-MKGpyIR-3P_NgWiz4Tz:bart20e) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/render/mod.rs#L207)):*
<details><summary><i>Previously, quodlibetor (Brandon W Maister) wrote…</i></summary><blockquote>

tioli: does this need to be a `clone()`? I suppose because it's wrapped in two non-`move` closures? I'm not sure if it would be possible to just inline the `get("materialized")` call here?
</blockquote></details>

It's in a loop so it does need a clone. Clones are reasonably cheap since the heavy stuff in the logger's behind an Rc. Inlining a call to get is the same but with a little additional overhead (it's basically a shim around the lookup and a clone).

Some of the threading boilerplate could be removed by calling get() from inside of source/mod.rs and passing the logger as a parameter to each record_offset_received/ingested call. It seems equally ugly to me so I have no preference, but let me know if it's preferable. 
___
*[src/dataflow/src/source/mod.rs, line 768 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoh-r1-768:-MKGu-jKEZ6BGaPipd63:b-1lqcc6) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/source/mod.rs#L768)):*
<details><summary><i>Previously, quodlibetor (Brandon W Maister) wrote…</i></summary><blockquote>

4 tiny allocations are worse than 1, though, aren't they?

I thought that the rowpacker had an internal freelist of rows that it could use for its allocations, reducing the likelihood of going all the way out to malloc.

I agree that moving this out so that it's not done in a batch is important enough to do either way.
</blockquote></details>

I saw no way to golf those allocs down because they eventually end up as keys in timely so need to conform to ExchangeData and be serializable via abomonation. If there's some form of owning, partially-built row type that's single alloc and can be appended to then I could use that as a key, but I didn't see anything along those lines.

I'll rearrange to batch these though
___
*[src/dataflow-types/src/logging.rs, line 141 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceob-r1-141:-MKGy4aN83GIlvxYehdd:bcz2kax) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow-types/src/logging.rs#L141)):*
<details><summary><i>Previously, quodlibetor (Brandon W Maister) wrote…</i></summary><blockquote>

These terms are very kafka-specific. I think that that makes sense -- trying to unify sources of different types in prometheus has been difficult/nonsensical -- but this table should probably be named `KafkaSourceInfo`.

IMO we should only add a higher-level "all sources" view on this once we have a unified theory about how to translate the different domains.
</blockquote></details>

These didn't seem kafka specific to me, but maybe I'm missing something? Offset seems like it'd be applicable to any sources we ingest, and file/kinesis both seem to generate them today based on a seen message counter.

**Review by quodlibetor — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4572)* status: 4 of 10 files reviewed, 7 unresolved discussions (waiting on @benesch, @elindsey, and @quodlibetor)

___
*[src/dataflow/src/source/mod.rs, line 768 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoh-r1-768:-MKH2PhY7Kll_IT-cZE_:brxp9sq) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/source/mod.rs#L768)):*
<details><summary><i>Previously, elindsey (Eli Lindsey) wrote…</i></summary><blockquote>

I saw no way to golf those allocs down because they eventually end up as keys in timely so need to conform to ExchangeData and be serializable via abomonation. If there's some form of owning, partially-built row type that's single alloc and can be appended to then I could use that as a key, but I didn't see anything along those lines.

I'll rearrange to batch these though
</blockquote></details>

Sounds good.
___
*[src/dataflow-types/src/logging.rs, line 141 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceob-r1-141:-MKH1Dao5PbR-o-CFPxo:bg69rdd) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow-types/src/logging.rs#L141)):*
<details><summary><i>Previously, elindsey (Eli Lindsey) wrote…</i></summary><blockquote>

These didn't seem kafka specific to me, but maybe I'm missing something? Offset seems like it'd be applicable to any sources we ingest, and file/kinesis both seem to generate them today based on a seen message counter.
</blockquote></details>

I think `partition_id` doesn't make any sense outside of kafka, although I suppose kinesis has `shards`, and offset can mean something different in different places (file offset in file, sequence number in kinesis) and I suppose we can just map everything to Kafka terminology, and default things to something if they really don't make sense.

I suppose using the Kafka terminology isn't any worse than us inventing our own, and yeah maybe I was thinking of something else, so this looks fine.

**Review by benesch — COMMENTED** [review]
> Reviewed 7 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4572)* status: all files reviewed, 7 unresolved discussions (waiting on @elindsey and @quodlibetor)

___
*[src/dataflow/src/render/mod.rs, line 207 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceod-r1-207:-MKI-lX1DdgLQhOEpsK4:b67uhco) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/render/mod.rs#L207)):*
<details><summary><i>Previously, elindsey (Eli Lindsey) wrote…</i></summary><blockquote>

It's in a loop so it does need a clone. Clones are reasonably cheap since the heavy stuff in the logger's behind an Rc. Inlining a call to get is the same but with a little additional overhead (it's basically a shim around the lookup and a clone).

Some of the threading boilerplate could be removed by calling get() from inside of source/mod.rs and passing the logger as a parameter to each record_offset_received/ingested call. It seems equally ugly to me so I have no preference, but let me know if it's preferable. 
</blockquote></details>

Cloning an `Rc` seems totally fine to me.
___
*[src/dataflow/src/source/mod.rs, line 768 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoh-r1-768:-MKHxjJI9TgVRRDDvk9P:bnb5clv) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/source/mod.rs#L768)):*
> 4 tiny allocations are worse than 1, though, aren't they?

Yeah, but I doubt it's expensive enough (now that the batching is in place) to justify the additional complexity of an Arc or a smol_str. Maybe someday, and definitely sooner rather than later if it shows up on performance profiles! But for now I suspect there is much lower-hanging performance fruit that it's not worth worrying about a few logging allocations, as long as they happen per-operator-invocation and not per-row.

> I thought that the rowpacker had an internal freelist of rows that it could use for its allocations, reducing the likelihood of going all the way out to malloc.

It does not. You can reuse a row packer over and over, to avoid allocating a new packer for every row, but that's the extent of it. IME building your own free list to avoid calling malloc doesn't typically make things go any faster. Jemalloc is really good at maintaining and searching free lists. You usually need very substantial information about allocation patterns to outperform jemalloc, and the row packer doesn't have any particular insight like that.

**Review by elindsey — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4572)* status: 5 of 10 files reviewed, 6 unresolved discussions (waiting on @benesch, @elindsey, and @quodlibetor)

___
*[src/coord/src/catalog/builtin.rs, line 98 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoa-r1-98:-MK_Gec56WhnhC3IrySb:b-896fja) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/coord/src/catalog/builtin.rs#L98)):*
<details><summary><i>Previously, elindsey (Eli Lindsey) wrote…</i></summary><blockquote>

Well, fixed half the original typo 🙂 
</blockquote></details>

Done!
___
*[src/dataflow/src/source/mod.rs, line 768 at r1](https://reviewable.io/reviews/materializeinc/materialize/4572#-MKBHficKhkfY_Cgceoh-r1-768:-MK_qlKq4rxfn65KaWtj:bivfmnr) ([raw file](https://github.com/materializeinc/materialize/blob/7baf113d9989bd0f7cada9ee25406fafe94938d5/src/dataflow/src/source/mod.rs#L768)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

> 4 tiny allocations are worse than 1, though, aren't they?

Yeah, but I doubt it's expensive enough (now that the batching is in place) to justify the additional complexity of an Arc or a smol_str. Maybe someday, and definitely sooner rather than later if it shows up on performance profiles! But for now I suspect there is much lower-hanging performance fruit that it's not worth worrying about a few logging allocations, as long as they happen per-operator-invocation and not per-row.

> I thought that the rowpacker had an internal freelist of rows that it could use for its allocations, reducing the likelihood of going all the way out to malloc.

It does not. You can reuse a row packer over and over, to avoid allocating a new packer for every row, but that's the extent of it. IME building your own free list to avoid calling malloc doesn't typically make things go any faster. Jemalloc is really good at maintaining and searching free lists. You usually need very substantial information about allocation patterns to outperform jemalloc, and the row packer doesn't have any particular insight like that.
</blockquote></details>

This now batches per operator invocation. Tracking both ingested and received doesn't make sense in that case since they'll be the same, so I've switched it to tracking a single offset in the system table. I left the prometheus metrics as-is so they still separately track received/ingested.

**Review by benesch — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4572#-:-MKcX0ge-L34FkZMZWdy:bnfp4nl)* status: all files reviewed, 6 unresolved discussions (waiting on @elindsey, @quodlibetor, and @umanwizard)
